src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,375 @@
+    }
+  };
+
+  // The serialized JS value and a bool to discern whether we've thrown an exception.
+  // Serializing may throw if the data we are serializing is not structured-cloneable.
+  // In this case, we will just serialize the exception message and set `caughtException`.
+  struct SerializedResult {
+    kj::Array<kj::byte> data;
+    bool caughtException;
+  };
+
+  // Serializes the return value of our method.
+  SerializedResult serialize(jsg::Lock& js, jsg::JsValue value) {
+    kj::Maybe<SerializedResult> result;
+
+    js.tryCatch([&] {
+      auto data = _::serializeV8(js, value.addRef(js).getHandle(js));
+      // Serialized okay!
+      result = SerializedResult { .data = kj::mv(data), .caughtException = false };
+    }, [this, &js, &result](jsg::Value exception) {
+      auto fullError = kj::str(jsg::check(exception.getHandle(js)->ToString(js.v8Context())));
+
+      // We may need to slice part of the message so we don't have `Error: Error: msg`.
+      kj::Maybe<kj::StringPtr> errorMessage(fullError);
+      if (fullError.startsWith("Error: "_kj)) {
+        errorMessage = fullError.slice(kj::size("Error: "_kj));
+      }
+
+      // Failed to serialize data.
+      auto failedResult = serialize(js, js.error(KJ_REQUIRE_NONNULL(errorMessage)));
+      // We are in the error branch, so we must set caughtException.
+      failedResult.caughtException = true;
+      result = kj::mv(failedResult);
+    });
+    return kj::mv(KJ_REQUIRE_NONNULL(result));
   }
+
+  // Invokes the given (synchronous) function, and serializes the result for RPC.
+  SerializedResult prepareSyncResult(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    kj::Maybe<v8::Local<v8::Value>> result;
+    bool caughtException = false;
+    js.tryCatch([&] {
+      result = invokeFn(js, fn, serializedArgs);
+    }, [&](jsg::Value exception) {
+      // We threw while executing JS.
+      caughtException = true;
+      result = exception.getHandle(js);
+    });
+
+    auto jsVal = jsg::JsValue(KJ_REQUIRE_NONNULL(result));
+    auto serializedResult = serialize(js, kj::mv(jsVal));
+
+    // `caughtException` might be true if the JS method threw an exception, so we don't want to
+    // accidentally set it to false just because serialization succeeded.
+    if (caughtException) {
+      serializedResult.caughtException = caughtException;
+    }
+    return kj::mv(serializedResult);
+  }
+
+  // Invokes the given (asynchronous) function, and serializes the result for RPC.
+  kj::Promise<SerializedResult> prepareAsyncResult(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    // We can't co_await a jsg::Promise, so we will use a promise-fulfiller pair instead.
+    auto [completionPromise, fulfiller] = kj::newPromiseAndFulfiller<SerializedResult>();
+    auto result = invokeFn(js, fn, serializedArgs);
```

Comment:
Keep in mind async functions can still throw synchronously when invoked. You need a trycatch here. (But per above, the sync and async paths need to be merged anyway.)